src/cpp/session/SessionSourceDatabaseSupervisor.cpp

```diff
@@ -116,8 +116,21 @@ FilePath sessionLockFilePath(const FilePath& sessionDir)
    return sessionDir.complete("lock_file");
 }
 
-// session dir lock (initialized by attachToSourceDatabase)
-FileLock s_sessionDirLock;
+// session dir lock (lock is acquired within 'attachToSourceDatabase()')
+boost::shared_ptr<FileLock> createSessionDirLock()
+{
+#ifdef RSTUDIO_SERVER
+   return FileLock::create(FileLock::LOCKTYPE_LINKBASED);
+#else
+   return FileLock::createDefault();
+#endif
+}
+
+boost::shared_ptr<FileLock>& sessionDirLock()
+{
+   static boost::shared_ptr<FileLock> instance = createSessionDirLock();
```

Comment:
I'm guessing that we are safe to assume that the destructor of FileLock executes at program termination but let's just test to make sure (not sure if shared_ptr creates some different destructor behavior, not likely at all but worth checking).